pkg/scheduler/framework/parallelize/parallelism.go

```diff
@@ -51,15 +51,28 @@ func chunkSizeFor(n, parallelism int) int {
 	return s
 }
 
+// numWorkersForChunkSize returns number of workers (goroutines)
+// that will be created in workqueue.ParallelizeUntil
+// for given parallelism, pieces and chunkSize values.
+func numWorkersForChunkSize(parallelism, pieces, chunkSize int) int {
+	chunks := (pieces + chunkSize - 1) / chunkSize
+	if chunks < parallelism {
+		return chunks
+	}
+	return parallelism
+}
+
 // Until is a wrapper around workqueue.ParallelizeUntil to use in scheduling algorithms.
 // A given operation will be a label that is recorded in the goroutine metric.
 func (p Parallelizer) Until(ctx context.Context, pieces int, doWorkPiece workqueue.DoWorkPieceFunc, operation string) {
+	chunkSize := chunkSizeFor(pieces, p.parallelism)
+	workers := numWorkersForChunkSize(p.parallelism, pieces, chunkSize)
+
 	goroutinesMetric := metrics.Goroutines.WithLabelValues(operation)
-	withMetrics := func(piece int) {
-		goroutinesMetric.Inc()
-		doWorkPiece(piece)
-		goroutinesMetric.Dec()
-	}
+	// Calling single Add with workers' count is more efficient than calling Inc or Dec per each work piece.
+	// This approach improves performance of some plugins (affinity, topology spreading).
+	goroutinesMetric.Add(float64(workers))
```

Comment:
I try to understand the code again (Sorry, my math is not good 😂), but it always seems to be equal to pieces. I write a program to verify

```go
package main

import (
	"context"
	"fmt"
	"sync"
	"sync/atomic"
)

func main() {
	// Test cases to verify
	testCases := []struct {
		workers   int
		pieces    int
		chunkSize int
	}{
		{2, 5, 2},   // Example from our discussion
		{3, 6, 2},   // Another example
		{2, 10, 3},  // Uneven chunks
		{5, 3, 1},   // More workers than pieces
		{1, 100, 7}, // Large number of pieces
	}

	for _, tc := range testCases {
		var callCount atomic.Int64

		// Create a function that counts calls
		doWork := func(piece int) {
			callCount.Add(1)
		}

		// Run the parallelizer
		ParallelizeUntil(context.Background(), tc.workers, tc.pieces, doWork, WithChunkSize(tc.chunkSize))

		fmt.Printf("Workers: %d, Pieces: %d, ChunkSize: %d => Actual calls: %d (Expected: %d)\n",
			tc.workers, tc.pieces, tc.chunkSize, callCount.Load(), tc.pieces)
	}
}

// Copy of the original ParallelizeUntil and supporting code for testing
func ParallelizeUntil(ctx context.Context, workers, pieces int, doWorkPiece func(piece int), opts ...Options) {
	if pieces == 0 {
		return
	}
	o := options{}
	for _, opt := range opts {
		opt(&o)
	}
	chunkSize := o.chunkSize
	if chunkSize < 1 {
		chunkSize = 1
	}

	chunks := ceilDiv(pieces, chunkSize)
	toProcess := make(chan int, chunks)
	for i := 0; i < chunks; i++ {
		toProcess <- i
	}
	close(toProcess)

	var stop <-chan struct{}
	if ctx != nil {
		stop = ctx.Done()
	}
	if chunks < workers {
		workers = chunks
	}
	wg := sync.WaitGroup{}
	wg.Add(workers)
	for i := 0; i < workers; i++ {
		go func() {
			defer wg.Done()
			for chunk := range toProcess {
				start := chunk * chunkSize
				end := start + chunkSize
				if end > pieces {
					end = pieces
				}
				for p := start; p < end; p++ {
					select {
					case <-stop:
						return
					default:
						doWorkPiece(p)
					}
				}
			}
		}()
	}
	wg.Wait()
}

type options struct {
	chunkSize int
}

type Options func(*options)

func WithChunkSize(c int) func(*options) {
	return func(o *options) {
		o.chunkSize = c
	}
}

func ceilDiv(a, b int) int {
	return (a + b - 1) / b
}
```

Output is

```
Workers: 2, Pieces: 5, ChunkSize: 2 => Actual calls: 5 (Expected: 5)
Workers: 3, Pieces: 6, ChunkSize: 2 => Actual calls: 6 (Expected: 6)
Workers: 2, Pieces: 10, ChunkSize: 3 => Actual calls: 10 (Expected: 10)
Workers: 5, Pieces: 3, ChunkSize: 1 => Actual calls: 3 (Expected: 3)
Workers: 1, Pieces: 100, ChunkSize: 7 => Actual calls: 100 (Expected: 100)
```

Reply:
BTW, it is weird that `goroutinesMetric` didn't equal the number of actual workers although it named `goroutines metric`

Reply:
We have a few different values in `parallelizer.Until`:
- parallelism (named `workers` in `ParallelizeUntil`) - how many goroutines (at most) we want to dispatch
- pieces - how many times we want to call the `doWorkPiece` function
- chunkSize - how many work pieces are processed at once within a goroutine
- chunks - into how many chunks the pieces are split (`ceilDiv(pieces, chunkSize)`)
- actual workers - is equal to `min(parallelism, chunks)` - that's the return value of `numWorkersForChunkSize`

The `actual workers` is the number of goroutines dispatched in `ParallelizeUntil`. In your code indeed `callCount.Load() == tc.piece`, because we run the `doWorkPiece` for each work piece. It doesn't mean that the `Goroutines` metric is wrong - it is capped at `actual workers` value here, as it is incremented before each work piece and decremented right after. We cannot have more than `actual workers` work pieces running at once, so it (almost) correctly counts the number of running goroutines. Almost, because we increment and decrement the metric multiple times within a single goroutine, what doesn't make sense.

Reply:
It's obvious that `callCount` equals `pieces`, right? What are you trying to say here?

Reply:
> Almost, because we increment and decrement the metric multiple times within a single goroutine, what doesn't make sense.

Yep, this is where I feel the original metric is a bit weird. By right it should be increased outside of `for chunk := range toProcess {`

> It's obvious that callCount equals pieces, right? What are you trying to say here?

I was thinking that may adjust the original code to make it record metrics accurately. 

```
	withMetrics := func(piece int) {
		doWorkPiece(piece)
		goroutinesMetric.Dec()
	}
	goroutinesMetric.Add(float64(pieces))
```